File: other/Julia/007.jl
#=
    https://projecteuler.net/problem=7
    By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
    What is the 10,001st prime number
=#

function p7a()
    n = 3
    count = 1
    primes = [2]

    while count < 10_001
        all(n % p != 0 for p ∈ primes) ? (push!(primes, n); n += 1; count += 1) : (n += 1)
    end
    return (primes[end])
end

function p7b()
    n = 3
    count = 1
    primes = [2]

    while count < 10_001
        if all(n % p != 0 for p ∈ primes if p ≤ sqrt(n))
            push!(primes, n)
            count += 1
        end
        n += 2
    end
    return (primes[end])
end

File: C++/007.cpp
/**
 * <p>By listing the first six prime numbers: $2, 3, 5, 7, 11$, and $13$, we can
 * see that the $6$th prime is $13$.</p> <p>What is the $10\,001$st prime
 * number?</p>
 */

#include <iostream>
#include <vector>

int main() {
  // all primes, except 2, are odd
  // make a list of primes
  // check against the list -> add or move on
  // count to 10'001

  constexpr auto N = 10'001; // NOLINT
  std::vector primes{2};
  primes.reserve(N);

  for (auto i = 3; primes.size() < N; i += 2) {
    bool composite = false;
    for (const auto &prime : primes) {
      if (i % prime == 0) {
        composite = true;
        break;
      }
    }
    if (!composite) {
      primes.push_back(i);
    }
  }
  std::cout << primes.back() << "\n";
}